source/common/router/rds_impl.cc

```diff
@@ -180,8 +180,21 @@ RouteConfigProviderSharedPtr RdsFactoryImpl::createRdsRouteConfigProvider(
     Server::Configuration::ServerFactoryContext& factory_context, const std::string& stat_prefix,
     Init::Manager& init_manager, ProtoTraitsImpl& proto_traits,
     Rds::RouteConfigProviderManager& manager) {
+  envoy::extensions::filters::network::http_connection_manager::v3::Rds normalized_rds;
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.normalize_rds_provider_config")) {
```

Comment:
I think it makes more sense to have this functionality in [addDynamicProvider](https://github.com/envoyproxy/envoy/blob/868d371518cceaf867d13d567d42b1c39d999fe2/source/common/rds/route_config_provider_manager.cc#L78). This way other uses of the same function will also benefit from the fix.

Reply:
That was my initial idea, but [addDynamicProvider](https://github.com/envoyproxy/envoy/blob/868d371518cceaf867d13d567d42b1c39d999fe2/source/common/rds/route_config_provider_manager.cc#L78) takes `const Protobuf::Message& rds` as argument, so I have to either change the signature of `addDynamicProvider` or cast/convert the first parameter `rds` to `envoy::extensions::filters::network::http_connection_manager::v3::Rds` in order to access `initial_fetch_timeout` field.

But I agree with you that this would be better. Let me try it again. I was not super familiar with that code when I tried it first time, so maybe now it will be easier to do.

Reply:
Moved the "normalization" to addDynamicProvider. I had to use reflection to find the right fields. The only assumption is that whatever dynamic config is passed, it must contain `config_source` field. If it does not, runtime assert will fail. Fortunately all components which use dynamic rds providers (generic_proxy, thrift_proxy and dubbo_proxy) have `config_source`. New components added in the future will have to use the same schema. I added a long comment next to assertion, so when it fails it will be easy to correct.